pkg/code-generator/codegen/templates/snapshot_template.go

```diff
@@ -80,6 +83,85 @@ func (s {{ .GoName }}Snapshot) HashFields() []zap.Field {
 	return append(fields, zap.Uint64("snapshotHash",  snapshotHash))
 }
 
+func (s *{{ .GoName }}Snapshot) GetResourcesList(resource resources.Resource) (resources.ResourceList, error) {
+	switch resource.(type) {
+{{- range .Resources }}
+	case *{{ .ImportPrefix }}{{ .Name }}:
+		return s.{{ upper_camel .PluralName }}.AsResources(), nil
+{{- end }}
+	default:
+		return resources.ResourceList{}, eris.New("did not contain the input resource type returning empty list")
+	}
+}
+
+func (s *{{ .GoName }}Snapshot) RemoveFromResourceList(resource resources.Resource) error {
+	refKey := resource.GetMetadata().Ref().Key()
+	switch resource.(type) {
+{{- range .Resources }}
+	case *{{ .ImportPrefix }}{{ .Name }}:
+		newList := {{ .ImportPrefix }}{{ .Name }}List{}
+		for _, res := range s.{{ upper_camel .PluralName }} {
+			if refKey != res.GetMetadata().Ref().Key() {
+				newList = append(newList, res)
+			}
+		}
+		s.{{ upper_camel .PluralName }} = newList
+		s.{{ upper_camel .PluralName }}.Sort()
+		return nil	
+{{- end }}
+	default:
+		return eris.Errorf("did not remove the reousource because its type does not exist [%T]", resource)
+	}
+}
+
+func (s *{{ .GoName }}Snapshot) AddOrReplaceToResourceList(resource resources.Resource) error {
+	refKey := resource.GetMetadata().Ref().Key() 
+	switch typed := resource.(type) {
+{{- range .Resources }}
+	case *{{ .ImportPrefix }}{{ .Name }}:
+		updated := false
+		for i, res := range s.{{ upper_camel .PluralName }} {
+			if refKey == res.GetMetadata().Ref().Key() {
+				s.{{ upper_camel .PluralName }}[i] = typed
+				updated = true
+			}
+		}
+		if !updated {
+			s.{{ upper_camel .PluralName }} = append(s.{{ upper_camel .PluralName }}, typed)
+		}
+		s.{{ upper_camel .PluralName }}.Sort()
+		return nil	
+{{- end }}
+	default:
+		return eris.Errorf("did not add/replace the resource type because it does not exist %T", resource)
+	}
+}
+
+func (s *{{ .GoName }}Snapshot) AddToResourceList(resource resources.Resource) error {
```

Comment:
why would we ever want this as opposed to the upsert? this would allow us to add duplicate resources to a snapshot which seems always wrong

Reply:
yeah sounds good